File: packages/core/ios/AgentConfigurationBuilder.swift
//
// Copyright 2025 Splunk Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import Foundation
import SplunkAgent

/// Builds native AgentConfiguration from the TS API.
enum AgentConfigurationBuilder {

  static func build(from dict: NSDictionary) throws -> AgentConfiguration {
    guard let endpointAny = dict.object(forKey: "endpoint"), !(endpointAny is NSNull) else {
      throw NSError(domain: "splunk.rn", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing endpoint in configuration"])
    }
    let endpoint = try buildEndpoint(from: endpointAny as? NSDictionary ?? NSDictionary())

    guard let appName = dict.object(forKey: "appName") as? String,
          let env = dict.object(forKey: "deploymentEnvironment") as? String else {
      throw NSError(domain: "splunk.rn", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing appName or deploymentEnvironment"])
    }

    var config = AgentConfiguration(endpoint: endpoint, appName: appName, deploymentEnvironment: env)
    if let appVersion = dict.object(forKey: "appVersion") as? String {
      config.appVersion = appVersion
    }

    if let debug = dict.object(forKey: "enableDebugLogging") as? Bool {
      config.enableDebugLogging = debug
    }

    if let global = dict.object(forKey: "globalAttributes") as? NSDictionary {
      config.globalAttributes = AttributeConverter.buildAttributes(from: global)
    }

    if let user = dict.object(forKey: "user") as? NSDictionary {
      var u = UserConfiguration()
      if let modeStr = user.object(forKey: "trackingMode") as? String {
        u.trackingMode = UserSessionHandler.deserializeUserTrackingMode(modeStr)
      }

      config.user = u
    }

    if let session = dict.object(forKey: "session") as? NSDictionary {
      var s = SessionConfiguration()

      if let samplingRate = session.object(forKey: "samplingRate") as? NSNumber {
        s.samplingRate = samplingRate.doubleValue
      } else if let samplingRate = session.object(forKey: "samplingRate") as? Double {
        s.samplingRate = samplingRate
      }

      config.session = s
    }

    return config
  }

  static func buildEndpoint(from dict: NSDictionary) throws -> EndpointConfiguration {
    if let realm = dict.object(forKey: "realm") as? String,
       let token = dict.object(forKey: "rumAccessToken") as? String {
      return EndpointConfiguration(realm: realm, rumAccessToken: token)
    }

    let tracesStr = dict.object(forKey: "trace") as? String
    let srStr = dict.object(forKey: "sessionReplay") as? String

    guard let tracesStr, let tracesURL = URL(string: tracesStr) else {
      throw NSError(domain: "splunk.rn", code: 2, userInfo: [NSLocalizedDescriptionKey: "Invalid trace"])
    }

    let srURL = srStr.flatMap { URL(string: $0) }
    return EndpointConfiguration(trace: tracesURL, sessionReplay: srURL)
  }
}

File: packages/core/android/src/main/kotlin/com/splunk/otel/reactnative/AgentConfigurationBuilder.kt
/*
 * Copyright 2025 Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.splunk.otel.reactnative

import com.facebook.react.bridge.ReadableMap
import com.splunk.rum.integration.agent.api.AgentConfiguration
import com.splunk.rum.integration.agent.api.EndpointConfiguration
import com.splunk.rum.integration.agent.api.user.UserTrackingMode
import io.opentelemetry.api.common.Attributes
import java.net.URL

/**
 * Builds native AgentConfiguration from the TS API.
 */
object AgentConfigurationBuilder {

  fun build(map: ReadableMap): AgentConfiguration {
    val endpointMapAny = map.getMap("endpoint")
      ?: throw IllegalArgumentException("endpoint is required")
    val endpointMap = endpointMapAny as ReadableMap

    val endpoint = if (endpointMap.hasKey("realm")) {
      EndpointConfiguration(endpointMap.getString("realm")!!, endpointMap.getString("rumAccessToken")!!)
    } else if (endpointMap.hasKey("trace")) {
      val traceStr = endpointMap.getString("trace")
        ?: throw IllegalArgumentException("trace must be a non-null string")
      val trace = URL(traceStr)

      val srStr = if (endpointMap.hasKey("sessionReplay")) endpointMap.getString("sessionReplay") else null

      val sr = srStr?.let { URL(it) }
      if (sr != null) EndpointConfiguration(trace, sr) else EndpointConfiguration(trace)
    } else {
      throw IllegalArgumentException("endpoint must specify either realm/rumAccessToken or trace")
    }

    val appName = map.getString("appName") ?: ""
    val deployment = map.getString("deploymentEnvironment") ?: ""
    val appVersion = if (map.hasKey("appVersion")) map.getString("appVersion") else null
    val enableDebug = map.getBoolean("enableDebugLogging")
    val global = if (map.hasKey("globalAttributes")) AttributeConverter.buildAttributesFromMap(map.getMap("globalAttributes")!!) else Attributes.empty()
    val instrumentedProcessName = if (map.hasKey("instrumentedProcessName")) map.getString("instrumentedProcessName") else null
    val deferredUntilForeground = map.getBoolean("deferredUntilForeground")

    val userMap = if (map.hasKey("user")) map.getMap("user") else null
    val userConfig = com.splunk.rum.integration.agent.api.user.UserConfiguration(
      trackingMode = userMap?.getString("trackingMode")?.let { UserTrackingMode.valueOf(it) } ?: UserTrackingMode.NO_TRACKING
    )

    val sessionMap = if (map.hasKey("session")) map.getMap("session") else null
    val sessionConfig = com.splunk.rum.integration.agent.api.session.SessionConfiguration(
      samplingRate = sessionMap?.getDouble("samplingRate") ?: 1.0
    )

    return AgentConfiguration(
      endpoint = endpoint,
      appName = appName,
      deploymentEnvironment = deployment,
      appVersion = appVersion,
      enableDebugLogging = enableDebug,
      globalAttributes = global,
      user = userConfig,
      session = sessionConfig,
      instrumentedProcessName = instrumentedProcessName,
      deferredUntilForeground = deferredUntilForeground,
    )
  }
}
